### ISSUE TITLE
Question on Lab 4 Level 5

### ISSUE THREAD

**Issue by chuashiqi** [nus-cs2030/2021-s1#84]
# Description 
For Lab 4 level 5, when asked to implement a toArray method, putting this: `public <T> T[] toArray(T[] t)` as my method header will return an error that says "The type parameter T is hiding the type T" but putting this: `public <U> U[] toArray(U[] u)` as my method header will not. 
Can anybody help to explain why changing the character used to represent the type will remove the error? Thanks!

## Topic: 
Generics

**Comment by marcuspeh** [comment]
I believe its because <T> was already declared in the class

**Comment by tlylt** [comment]
Given that you likely used `class ImmutableList<T> {` in your class declaration, `T` already has a meaning in your class body.
If you do 
```java
public <U> U[] toArray(U[] u){
```
You are creating a new type parameter `U` which has nothing to do with `T`, and `U` is within the scope of the method `toArray()`
However, if you do
```java
public <T> T[] toArray(T[] t){
```
`T` already has a meaning in this class, and now you want it to be another type parameter. So it is kind of like having two type parameters but they have the same name. This is kind of ambiguous.

I wanted to conclude that hence we cannot use the same letter, but I went to try using `public <T> T[] toArray(T[] t)` and it compiles without warning and I managed to run test5.jsh without a problem as well. So, are you getting it as an error(or warning) during compilation or when you are running test5? Or both?

**Comment by waynetee** [comment]
@shhiiiqqqqiiiii its a little bit tricky to debug without any code to look at, you don't happen to have "T" declared anywhere else right? (Other than `class ImmutableList<T>`)  and you aren't overriding the method right?

**Comment by chuashiqi** [comment]
@tlylt I just tried compiling using `public <T> T[] toArray(T[] t)` and actually get no error when compiling but there's a warning message that pops up in my vim terminal:
<img width="445" alt="Untitled" src="https://user-images.githubusercontent.com/70755428/94016229-707d8f80-fde0-11ea-9cba-fae72a238fc6.png">
 which was why I just changed to U and never tried compiling with T. I was just curious why I had a warning pop up when even the prof uses declares T for both class and method. 

Your explanation does make sense to me though so thanks for that!
